components/camel-vertx-kafka/camel-vertx-kafka-component/src/main/java/org/apache/camel/component/vertx/kafka/VertxKafkaConsumer.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.camel.ExtendedExchange;
+import org.apache.camel.Processor;
+import org.apache.camel.component.vertx.kafka.configuration.VertxKafkaConfiguration;
+import org.apache.camel.component.vertx.kafka.operations.VertxKafkaConsumerOperations;
+import org.apache.camel.spi.Synchronization;
+import org.apache.camel.support.DefaultConsumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class VertxKafkaConsumer extends DefaultConsumer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(VertxKafkaConsumer.class);
+
+    private KafkaConsumer<Object, Object> kafkaConsumer;
+
+    public VertxKafkaConsumer(final VertxKafkaEndpoint endpoint, final Processor processor) {
+        super(endpoint, processor);
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+
+        // create the consumer client
+        kafkaConsumer = KafkaConsumer.create(getEndpoint().getVertx(), getConfiguration().createConsumerConfiguration());
+
+        // create the consumer operation
+        final VertxKafkaConsumerOperations consumerOperations
+                = new VertxKafkaConsumerOperations(kafkaConsumer, getConfiguration());
+
+        // process our records
+        consumerOperations.receiveEvents(this::onEventListener, this::onErrorListener);
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        if (kafkaConsumer != null) {
+            kafkaConsumer.close();
+        }
+
+        super.doStop();
+    }
+
+    public VertxKafkaConfiguration getConfiguration() {
+        return getEndpoint().getConfiguration();
+    }
+
+    @Override
+    public VertxKafkaEndpoint getEndpoint() {
+        return (VertxKafkaEndpoint) super.getEndpoint();
+    }
+
+    private void onEventListener(final KafkaConsumerRecord<Object, Object> record) {
+        final Exchange exchange = getEndpoint().createVertxKafkaExchange(record);
+        // set propagated headers
+        VertxKafkaHeadersPropagation.getPropagatedHeaders(record.headers(), exchange.getIn())
+                .forEach((key, value) -> exchange.getIn().setHeader(key, value));
+
+        // add exchange callback
+        exchange.adapt(ExtendedExchange.class).addOnCompletion(new Synchronization() {
+            @Override
+            public void onComplete(Exchange exchange) {
+                // at the moment we don't commit the offsets manually, we can add it in the future
+            }
+
+            @Override
+            public void onFailure(Exchange exchange) {
+                // we do nothing here
+                processRollback(exchange);
+            }
+        });
+        // send message to next processor in the route
+        getAsyncProcessor().process(exchange, doneSync -> LOG.trace("Processing exchange [{}] done.", exchange));
+    }
+
+    private void onErrorListener(final Throwable error) {
+        final Exchange exchange = getEndpoint().createExchange();
```

Comment:
Looks like here you need to set the exception on the created exchange as otherwise its null

Reply:
Ah good note. Didnt realize that it will be always null